### DISCUSSION TITLE
Implement a way to create data without automatic charmap translation

### DISCUSSION THREAD

**Discussion by ReallyNiceGuy** [cc65/cc65#1468] [Ideas]
My suggestion:

Just implement a single new command: .RBYTE, which will bypass any translation in charmap and use the code points as they are in the source code file.
This would simplify coding in mixed charset mode in Commander X16.

**Comment by oliverschmidt** [comment]
I must admit that I'm not "happy" with `.RBYTE`.

Maybe it's just me, but .BYTE is already sort of strange. With the term "byte" I associate `.byte $10, $65` but not `.byte "Hello"`.

And now we would have .RBYTE. And for what I consider "normal", it wouldn't make a difference at all: `.rbyte $10, $65` would be just the same as `.byte $10, $65`. Not intuitive at all - at least to me.

I'm wondering if this would make sense:

* Introduce `.ASCII` just like `.ASCIIZ` without the terminator.
* Add a `.FEATURE` to allow to control if the arguments of `.BYTE` should be subject to char mapping or not. For backward compatibility the feature would be turned on by default. New code could choose between
  * Using .feature to switch between both types of .byte
  * Using .feature once to turn .byte mapping off and use .ascii if mapping is desired.

**Reply by ReallyNiceGuy** [reply]
> I must admit that I'm not "happy" with `.RBYTE`.
> 
> Maybe it's just me, but .BYTE is already sort of strange. With the term "byte" I associate `.byte $10, $65` but not `.byte "Hello"`.

I agree, but because of another reason: .ASCIIZ is superfluous.
You can do `.BYTE "Hello",0` It even uses less characters than `.ASCIIZ "Hello"` ;)
Also, .ASCIIZ is a misnomer, as it will do translation to the current charmap and not, as I would expect, leave the string as ascii with a null byte at the end.

> 
> And now we would have .RBYTE. And for what I consider "normal", it wouldn't make a difference at all: `.rbyte $10, $65` would be just the same as `.byte $10, $65`. Not intuitive at all - at least to me.

That is a failure of the .BYTE keyword, which behaves completely differently if you use a character literal or a value literal.

.RBYTE would be consistent from the start. .RBYTE 'a' would be the same as .RBYTE $61, as long as you have your source code using ascii or latin1. It won't even compile in any other encoding, due to assumptions on the compiler itself.

> 
> I'm wondering if this would make sense:
> 
> * Introduce `.ASCII` just like `.ASCIIZ` without the terminator.

With .RBYTE (I don't have a strong preference on the real name of the keyword, but ASCII is confusing, as it implies that it is the same as ASCIIZ without the terminating null byte, but it definitely is not).

> * Add a `.FEATURE` to allow to control if the arguments of `.BYTE` should be subject to char mapping or not. For backward compatibility the feature would be turned on by default. New code could choose between
>   
>   * Using .feature to switch between both types of .byte
>   * Using .feature once to turn .byte mapping off and use .ascii if mapping is desired.

This switcheroo will create the same problem that we tried to solve with pushcharmp and popcharmap. How to go back to the way the .feature was defined before our code was included?

**Comment by oliverschmidt** [comment]
> That is a failure of the .BYTE keyword, which behaves completely differently if you use a character literal or a value literal.

That's right.

> .RBYTE would be consistent from the start. .RBYTE 'a' would be the same as .RBYTE $61, as long as you have your source code using ascii or latin1. It won't even compile in any other encoding, due to assumptions on the compiler itself.

That's of course technically right. But I have a problem with the name. "raw byte" sounds like a special byte. But in fact it's a "less special" byte. And given that - as we both agree - .byte is already strange...

You see: I'd like to get to the point where we reduce complexity / strangeness instead of adding another layer on top of it - and even if it's only the name. Or to put it in other words: Either we improve .byte in one way or another or we leave it alone (incl. the name "byte").

> This switcheroo will create the same problem that we tried to solve with pushcharmp and popcharmap. How to go back to the way the .feature was defined before our code was included?

I see. Then maybe we don't allow to switch it for a single source file at all. Maybe just a cmdline switch to provide backward compatibility. Without the switch .byte behaves like your .rbyte.

**Reply by ReallyNiceGuy** [reply]
> > That is a failure of the .BYTE keyword, which behaves completely differently if you use a character literal or a value literal.
> 
> That's right.
> 
> > .RBYTE would be consistent from the start. .RBYTE 'a' would be the same as .RBYTE $61, as long as you have your source code using ascii or latin1. It won't even compile in any other encoding, due to assumptions on the compiler itself.
> 
> That's of course technically right. But I have a problem with the name. "raw byte" sounds like a special byte. But in fact it's a "less special" byte. And given that - as we both agree - .byte is already strange...
> 
> You see: I'd like to get to the point where we reduce complexity / strangeness instead of adding another layer on top of it - and even if it's only the name. Or to put it in other words: Either we improve .byte in one way or another or we leave it alone (incl. the name "byte").
> 

We leave it alone.

> > This switcheroo will create the same problem that we tried to solve with pushcharmp and popcharmap. How to go back to the way the .feature was defined before our code was included?
> 
> I see. Then maybe we don't allow to switch it for a single source file at all. Maybe just a cmdline switch to provide backward compatibility. Without the switch .byte behaves like your .rbyte.

This is even worse. Change .BYTE behavior with a command line switch is just shifting the issue one level up.
How would I even create translated strings if I use this magical switch?

What about .LITERAL ?
We want to have a way to pass literal strings to the compiler to be embedded in the object.

**Comment by oliverschmidt** [comment]
> How would I even create translated strings if I use this magical switch?

By using .ASCII.

**Reply by ReallyNiceGuy** [reply]
> > How would I even create translated strings if I use this magical switch?
> 
> By using .ASCII.

I don't like breaking backwards compatibility. If you are going to create a new keyword, at least leave the old keywords behaving as they always had. Broken or not, it is what was specified and what people are using.

**Comment by oliverschmidt** [comment]
Okay - I'm fine with .LITERAL (instead of .RBYTE).

**Reply by ReallyNiceGuy** [reply]
Great, I will start with this change after I finish the current one.

Cheers!
